File: src/main/java/net/explorviz/adapter/service/converter/AttributesReader.java
package net.explorviz.adapter.service.converter;

import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_APP_INSTANCE_ID;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_APP_LANG;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_APP_NAME;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_CLASS_FQN;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_HOST_IP;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_HOST_NAME;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_LANDSCAPE_SECRET;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_LANDSCAPE_TOKEN;
import static net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_METHOD;

import io.opentelemetry.proto.common.v1.AnyValue;
import io.opentelemetry.proto.trace.v1.Span;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * Reads the attributes of a {@link Span}.
 */
public class AttributesReader {

  /**
   * The token that uniquely identifies the landscape a span belongs to.
   */
  public static final String LANDSCAPE_TOKEN = "landscape_token";

  /**
   * The token's secret.
   */
  public static final String TOKEN_SECRET = "token_secret";

  /**
   * The name of the host.
   */
  public static final String HOST_NAME = "host";

  /**
   * The IP address the application runs on.
   */
  public static final String HOST_IP = "host_address";

  /**
   * The name of the application a span belongs to.
   */
  public static final String APPLICATION_NAME = "application_name";


  /**
   * The instance id of the application.
   */
  public static final String APPLICATION_INSTANCE_ID = "application_instance_id";

  /**
   * The programming language that the application is written in.
   */
  public static final String APPLICATION_LANGUAGE = "application_language";

  /**
   * The identifier of the operation/method called.
   */
  public static final String CODE_FUNCTION = "code.function";

  /**
   * The fully qualified class name to which the called operation/method belongs.
   */
  public static final String CODE_NAMESPACE = "code.namespace";

  /**
   * The fully qualified name of the operation/method called.
   */
  public static final String METHOD_FQN = "java.fqn";

  /*
   * Default values
   */

  // k8s section

  public static final String K8S_POD_NAME = "k8s.pod.name";
  public static final String K8S_NAMESPACE_NAME = "k8s.namespace.name";
  public static final String K8S_NODE_NAME = "k8s.node.name";
  public static final String K8S_DEPLOYMENT_NAME = "k8s.deployment.name";


  private final Map<String, AnyValue> attributes = new HashMap<>(7);

  private Span span = null;

  /**
   * Reads attributes from a span.
   *
   * @param span the span to read attributes out of
   */
  public AttributesReader(final Span span) {
    this.span = span;

    // Load attributes into map
    span.getAttributesList().forEach(keyValue -> {
      attributes.put(keyValue.getKey(), keyValue.getValue());
    });
  }


  /**
   * Unwraps an AttributeValue of into a string.
   *
   * @param key the attribute's key
   * @return the string value of the attribute or empty if no such key exists
   */
  private Optional<String> getAsString(final String key) {
    final AnyValue av = this.attributes.get(key);
    if (av == null) {
      return Optional.empty();
    }

    return Optional.of(av.getStringValue());
  }

  public String getLandscapeToken() {
    return this.getAsString(LANDSCAPE_TOKEN).orElse(DEFAULT_LANDSCAPE_TOKEN);
  }

  public String getSecret() {
    return this.getAsString(TOKEN_SECRET).orElse(DEFAULT_LANDSCAPE_SECRET);
  }

  public String getHostName() {
    return this.getAsString(HOST_NAME).orElse(DEFAULT_HOST_NAME);
  }

  public String getHostIpAddress() {
    return this.getAsString(HOST_IP).orElse(DEFAULT_HOST_IP);
  }

  public String getApplicationName() {
    return this.getAsString(APPLICATION_NAME).orElse(DEFAULT_APP_NAME);
  }

  public String getApplicationInstanceId() {
    return this.getAsString(APPLICATION_INSTANCE_ID).orElse(DEFAULT_APP_INSTANCE_ID);
  }

  public String getApplicationLanguage() {
    return this.getAsString(APPLICATION_LANGUAGE).orElse(DEFAULT_APP_LANG);
  }

  public String getMethodFqn() {
    String fallbackMethodName = this.span.getName();
    if (fallbackMethodName.isEmpty()) {
      fallbackMethodName = DEFAULT_METHOD;
    }

    Optional<String> codeNamespace = this.getAsString(CODE_NAMESPACE);
    Optional<String> codeFunction = this.getAsString(CODE_FUNCTION);
    Optional<String> methodFqn = this.getAsString(METHOD_FQN);

    return codeNamespace.flatMap(classFqn -> codeFunction.map(methodName -> classFqn + "." + methodName))
        .or(() -> methodFqn)
        .orElse(DEFAULT_CLASS_FQN + fallbackMethodName);
  }

  public String getK8sPodName() {
    return this.getAsString(K8S_POD_NAME).orElse("");
  }

  public String getK8sNamespace() {
    return this.getAsString(K8S_NAMESPACE_NAME).orElse("");
  }

  public String getK8sNodeName() {
    return this.getAsString(K8S_NODE_NAME).orElse("");
  }

  public String getK8sDeploymentName() {
    return this.getAsString(K8S_DEPLOYMENT_NAME).orElse("");
  }


  /**
   * Appends all attributes to the given {{@link net.explorviz.avro.Span}} builder.
   *
   * @param builder the builder to append the attributes to
   */
  public void appendToSpan(final net.explorviz.avro.Span.Builder builder) {
    builder
        .setLandscapeToken(this.getLandscapeToken())
        .setHostname(this.getHostName())
        .setHostIpAddress(this.getHostIpAddress())
        .setAppInstanceId(this.getApplicationInstanceId())
        .setAppName(this.getApplicationName())
        .setAppLanguage(this.getApplicationLanguage())
        .setFullyQualifiedOperationName(this.getMethodFqn())
        .setK8sPodName(this.getK8sPodName())
        .setK8sNamespace(this.getK8sNamespace())
        .setK8sNodeName(this.getK8sNodeName())
        .setK8sDeploymentName(this.getK8sDeploymentName());
  }
}

File: src/main/java/net/explorviz/adapter/service/converter/AttributesReader.kt
package net.explorviz.adapter.service.converter

import io.opentelemetry.proto.common.v1.AnyValue
import io.opentelemetry.proto.trace.v1.Span
import net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_APP_NAME
import net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_FUNCTION_NAME
import net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_LANDSCAPE_SECRET
import net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_LANDSCAPE_TOKEN
import net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_NAMESPACE
import net.explorviz.adapter.service.converter.DefaultAttributeValues.DEFAULT_PACKAGE_NAME

/** Reads the attributes of a [Span]. */
open class AttributesReader(private val span: Span) {

    companion object {
        const val LANDSCAPE_TOKEN = "explorviz.token.id"
        const val TOKEN_SECRET = "explorviz.token.secret"
        const val GIT_COMMIT_CHECKSUM = "git_commit_checksum"
        const val HOST_NAME = "host"
        const val HOST_IP = "host_address"
        const val APPLICATION_NAME = "service.name"
        const val APPLICATION_INSTANCE_ID = "service.instance.id"
        const val APPLICATION_LANGUAGE = "telemetry.sdk.language"
        const val CODE_FUNCTION = "code.function"
        const val CODE_NAMESPACE = "code.namespace"
        const val METHOD_FQN = "java.fqn"
        const val K8S_POD_NAME = "k8s.pod.name"
        const val K8S_NAMESPACE_NAME = "k8s.namespace.name"
        const val K8S_NODE_NAME = "k8s.node.name"
        const val K8S_DEPLOYMENT_NAME = "k8s.deployment.name"
    }

    private val attributes: Map<String, AnyValue> = span.attributesList.associate { it.key to it.value }

    open fun getAsString(key: String): String? {
        return attributes[key]?.stringValue
    }

    val landscapeToken: String
        get() = getAsString(LANDSCAPE_TOKEN) ?: DEFAULT_LANDSCAPE_TOKEN

    val secret: String
        get() = getAsString(TOKEN_SECRET) ?: DEFAULT_LANDSCAPE_SECRET

    val hostName: String?
        get() = getAsString(HOST_NAME)

    val hostIpAddress: String?
        get() = getAsString(HOST_IP)

    val gitCommitChecksum: String?
        get() = getAsString(GIT_COMMIT_CHECKSUM)

    val applicationName: String
        get() = getAsString(APPLICATION_NAME) ?: DEFAULT_APP_NAME

    val applicationInstanceId: String?
        get() = getAsString(APPLICATION_INSTANCE_ID)

    val applicationLanguage: String?
        get() = getAsString(APPLICATION_LANGUAGE)

    val namespace: String
        get() {
            return getAsString(CODE_NAMESPACE) ?: generateNamespaceFromSpanName()
        }

    val functionName: String
        get() {
            return getAsString(CODE_FUNCTION) ?: generateFunctionNameFromSpanName()
        }

    open fun generateNamespaceFromSpanName(): String {
        val spanName = span.name
        if (spanName.isNullOrEmpty()) return DEFAULT_NAMESPACE

        val fqnComponents = spanName.split(".")

        val namespaceInSpanName =
            spanName.substring(
                0,
                Math.max(0, spanName.length - (fqnComponents.last().length + 1)),
            )

        return when (fqnComponents.size) {
            1 -> DEFAULT_NAMESPACE
            2 -> "${DEFAULT_PACKAGE_NAME}.${namespaceInSpanName}"
            else -> namespaceInSpanName
        }
    }

    open fun generateFunctionNameFromSpanName(): String {
        val spanName = span.name
        if (spanName.isNullOrEmpty()) return DEFAULT_FUNCTION_NAME

        val fqnComponents = spanName.split(".")

        return if (fqnComponents.size <= 1) {
            spanName
        } else {
            spanName.substring(
                spanName.length - fqnComponents.last().length,
                spanName.length,
            )
        }
    }

    val k8sPodName: String?
        get() = getAsString(K8S_POD_NAME)

    val k8sNamespace: String?
        get() = getAsString(K8S_NAMESPACE_NAME)

    val k8sNodeName: String?
        get() = getAsString(K8S_NODE_NAME)

    val k8sDeploymentName: String?
        get() = getAsString(K8S_DEPLOYMENT_NAME)

    fun appendToSpan(builder: net.explorviz.avro.Span.Builder) {
        builder.apply {
            landscapeToken = this@AttributesReader.landscapeToken
            appName = this@AttributesReader.applicationName
            namespace = this@AttributesReader.namespace
            functionName = this@AttributesReader.functionName
            gitCommitChecksum = this@AttributesReader.gitCommitChecksum
            hostname = this@AttributesReader.hostName
            hostIpAddress = this@AttributesReader.hostIpAddress
            appInstanceId = this@AttributesReader.applicationInstanceId
            appLanguage = this@AttributesReader.applicationLanguage
            k8sPodName = this@AttributesReader.k8sPodName
            k8sNamespace = this@AttributesReader.k8sNamespace
            k8sNodeName = this@AttributesReader.k8sNodeName
            k8sDeploymentName = this@AttributesReader.k8sDeploymentName
        }
    }
}
